// @sap-cloud-sdk/core/dist/odata/v2/request-builder/response-data-accessor.js
"use strict";
/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSingleResult = exports.isCollectionResult = exports.getCollectionResult = void 0;
var util_1 = require("@sap-cloud-sdk/util");
var logger = util_1.createLogger({
    package: 'core',
    messageContext: 'response-data-accessor'
});
function getCollectionResult(data) {
    var _a;
    validateCollectionResult(data);
    return isCollectionResult(data) ? (_a = data === null || data === void 0 ? void 0 : data.d) === null || _a === void 0 ? void 0 : _a.results : [];
}
exports.getCollectionResult = getCollectionResult;
function isCollectionResult(data) {
    var _a;
    return Array.isArray((_a = data === null || data === void 0 ? void 0 : data.d) === null || _a === void 0 ? void 0 : _a.results);
}
exports.isCollectionResult = isCollectionResult;
function validateCollectionResult(data) {
    if (!isCollectionResult(data)) {
        logger.warn('The given reponse data does not have the standard OData v2 format for collections.');
    }
}
function getSingleResult(data) {
    var _a;
    validateSingleResult(data);
    return isSingleResultAsCollection(data) ? (_a = data === null || data === void 0 ? void 0 : data.d) === null || _a === void 0 ? void 0 : _a.results : (data === null || data === void 0 ? void 0 : data.d) || {};
}
exports.getSingleResult = getSingleResult;
// Workaround to be compatible with services that wrongly implement the OData v2 protocol and serve single responses in the same format as collections
function isSingleResultAsCollection(data) {
    var _a;
    return !!((_a = data === null || data === void 0 ? void 0 : data.d) === null || _a === void 0 ? void 0 : _a.results) && !isCollectionResult(data);
}
function validateSingleResult(data) {
    if (isSingleResultAsCollection(data)) {
        logger.warn('The given reponse data has the format for collections instead of the standard OData v2 format for single results.');
    }
    if (!(data === null || data === void 0 ? void 0 : data.d)) {
        logger.warn('The given reponse data does not have the standard OData v2 format for single results.');
    }
}
