// @sap-cloud-sdk/core/dist/odata/common/request/odata-request.js
"use strict";
/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ODataRequest = void 0;
var util_1 = require("@sap-cloud-sdk/util");
var scp_cf_1 = require("../../../scp-cf");
var remove_slashes_1 = require("../../../util/remove-slashes");
var http_client_1 = require("../../../http-client");
var header_builder_1 = require("../../../header-builder");
// TODO: The buildCsrfHeaders import cannot be combined with the rest of the other headers due to circular dependencies
var csrf_token_header_1 = require("../../../header-builder/csrf-token-header");
// TODO: The buildHeadersForDestination import cannot be combined with the rest of the other headers due to circular dependencies
var header_builder_for_destination_1 = require("../../../header-builder/header-builder-for-destination");
var odata_request_traits_1 = require("./odata-request-traits");
/**
 * OData request configuration for an entity type.
 *
 * @typeparam EntityT - Type of the entity to setup a request for
 */
var ODataRequest = /** @class */ (function () {
    /**
     * Creates an instance of ODataRequest.
     *
     * @param config - Configuration of the request
     * @param _destination - Destination to setup the request against
     * @memberof ODataRequest
     */
    function ODataRequest(config, _destination) {
        this.config = config;
        this._destination = _destination;
    }
    Object.defineProperty(ODataRequest.prototype, "destination", {
        get: function () {
            return this._destination;
        },
        set: function (dest) {
            this._destination = dest && scp_cf_1.sanitizeDestination(dest);
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Constructs the url for the request at hand.
     *
     * @returns The url for the request
     */
    ODataRequest.prototype.url = function () {
        return "" + remove_slashes_1.removeTrailingSlashes(this.resourceUrl()) + this.query();
    };
    /**
     * Constructs the relative url for the batch request.
     *
     * @returns The relative url for the batch request
     */
    ODataRequest.prototype.relativeUrl = function () {
        return "" + remove_slashes_1.removeTrailingSlashes(this.relativeResourceUrl()) + this.query();
    };
    /**
     * Specifies whether the destination needs a specific authentication or not.
     *
     * @returns A boolean value that specifies whether the destination needs authentication or not
     * @memberof ODataRequest
     */
    ODataRequest.prototype.needsAuthentication = function () {
        return (!!this.destination &&
            this.destination.authentication !== 'NoAuthentication');
    };
    /**
     * Returns the service URL for a given OData request.
     *
     * @returns The URL of the service the given entity belongs to
     */
    ODataRequest.prototype.serviceUrl = function () {
        if (!this.destination) {
            throw Error('The destination is undefined.');
        }
        var systemUrl = this.destination.url;
        var servicePath = typeof this.config.customServicePath === 'undefined'
            ? this.config.defaultServicePath
            : this.config.customServicePath;
        return remove_slashes_1.removeTrailingSlashes(systemUrl) + "/" + remove_slashes_1.removeSlashes(servicePath);
    };
    /**
     * Returns the relative service URL for a given OData request.
     *
     * @returns The relative URL of the service the given entity belongs to
     */
    ODataRequest.prototype.relativeServiceUrl = function () {
        var servicePath = typeof this.config.customServicePath === 'undefined'
            ? this.config.defaultServicePath
            : this.config.customServicePath;
        return "" + remove_slashes_1.removeSlashes(servicePath);
    };
    /**
     * Returns the URL to a specific OData resource, i.e. the entity collection.
     *
     * @returns The URL of the resource
     */
    ODataRequest.prototype.resourceUrl = function () {
        return remove_slashes_1.removeTrailingSlashes(this.serviceUrl()) + "/" + this.config.resourcePath();
    };
    /**
     * Returns the relative URL to a specific OData resource.
     *
     * @returns The relative URL of the resource
     */
    ODataRequest.prototype.relativeResourceUrl = function () {
        return remove_slashes_1.removeTrailingSlashes(this.relativeServiceUrl()) + "/" + this.config.resourcePath();
    };
    /**
     * Get query parameters as string. Leads with `?` if  there are parameters to return.
     *
     * @returns Query parameter string
     */
    ODataRequest.prototype.query = function () {
        var queryParameters = __assign(__assign({}, this.config.queryParameters()), this.config.customQueryParameters);
        var query = Object.entries(queryParameters)
            .map(function (_a) {
            var key = _a[0], value = _a[1];
            return key + "=" + value;
        })
            .join('&');
        return query.length ? "?" + query : '';
    };
    /**
     * Create object containing all headers, including custom headers for the given request.
     *
     * @returns Key-value pairs where the key is the name of a header property and the value is the respective value
     */
    ODataRequest.prototype.headers = function () {
        return __awaiter(this, void 0, void 0, function () {
            var defaultHeaders, destinationRelatedHeaders, csrfHeaders, _a, error_1;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 5, , 6]);
                        if (!this.destination) {
                            throw Error('The destination is undefined.');
                        }
                        defaultHeaders = header_builder_1.replaceDuplicateKeys(header_builder_1.filterNullishValues(__assign({ accept: 'application/json', 'content-type': this.config.contentType }, this.getETagHeader())), this.config.customHeaders);
                        return [4 /*yield*/, header_builder_for_destination_1.buildHeadersForDestination(this.destination, this.config.customHeaders)];
                    case 1:
                        destinationRelatedHeaders = _b.sent();
                        if (!(this.config.method === 'get')) return [3 /*break*/, 2];
                        _a = {};
                        return [3 /*break*/, 4];
                    case 2: return [4 /*yield*/, this.getCsrfHeaders(destinationRelatedHeaders)];
                    case 3:
                        _a = _b.sent();
                        _b.label = 4;
                    case 4:
                        csrfHeaders = _a;
                        return [2 /*return*/, __assign(__assign(__assign(__assign({}, destinationRelatedHeaders), csrfHeaders), defaultHeaders), this.config.customHeaders)];
                    case 5:
                        error_1 = _b.sent();
                        return [2 /*return*/, Promise.reject(util_1.errorWithCause('Constructing headers for OData request failed!', error_1))];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Execute the given request and return the according promise.
     *
     * @returns Promise resolving to the requested data
     */
    ODataRequest.prototype.execute = function () {
        return __awaiter(this, void 0, void 0, function () {
            var destination, _a, _b, _c;
            var _this = this;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        destination = this.destination;
                        if (!destination) {
                            throw Error('The destination cannot be undefined.');
                        }
                        _a = http_client_1.executeHttpRequest;
                        _b = [destination];
                        _c = {};
                        return [4 /*yield*/, this.headers()];
                    case 1: return [2 /*return*/, _a.apply(void 0, _b.concat([(_c.headers = _d.sent(),
                                _c.url = this.relativeUrl(),
                                _c.method = this.config.method,
                                _c.data = this.config.payload,
                                _c)])).catch(function (error) {
                            return Promise.reject(constructError(error, _this.config.method, _this.serviceUrl()));
                        })];
                }
            });
        });
    };
    ODataRequest.prototype.getETagHeader = function () {
        var eTag = odata_request_traits_1.isWithETag(this.config)
            ? this.config.versionIdentifierIgnored
                ? '*'
                : this.config.eTag
            : undefined;
        return header_builder_1.filterNullishValues({ 'if-match': eTag });
    };
    ODataRequest.prototype.getCsrfHeaders = function (destinationRelatedHeaders) {
        return __awaiter(this, void 0, void 0, function () {
            var customCsrfHeaders;
            return __generator(this, function (_a) {
                customCsrfHeaders = header_builder_1.getHeader('x-csrf-token', this.config.customHeaders);
                return [2 /*return*/, Object.keys(customCsrfHeaders).length
                        ? customCsrfHeaders
                        : csrf_token_header_1.buildCsrfHeaders(this.destination, {
                            headers: destinationRelatedHeaders,
                            url: this.relativeServiceUrl()
                        })];
            });
        });
    };
    return ODataRequest;
}());
exports.ODataRequest = ODataRequest;
function constructError(error, requestMethod, url) {
    if (error.isAxiosError) {
        var defaultMessage = requestMethod + " request to " + url + " failed!";
        var s4SpecificMessage = util_1.propertyExists(error, 'response', 'data', 'error')
            ? messageFromS4ErrorResponse(error)
            : '';
        var message = [defaultMessage, s4SpecificMessage].join(' ');
        return util_1.errorWithCause(message, error);
    }
    return error;
}
function messageFromS4ErrorResponse(error) {
    return (util_1.propertyExists(error.response.data.error, 'message', 'value')
        ? error.response.data.error.message.value
        : '') + "\n" + JSON.stringify(error.response.data.error);
}
