// @sap/cds-runtime/lib/hana/pool.js
const cds = global.cds || require('@sap/cds/lib')

const pool = require('@sap/cds-foss')('generic-pool')
const hana = require('./hanaDriver')

const _require = require('../common/utils/require')

let im

function multiTenantInstanceManager (db = cds.env.requires.db) {
  const creds =
    (db.credentials && (db.credentials.get_managed_instance_url || db.credentials.sm_url) && db.credentials) ||
    _require('@sap/xsenv').serviceCredentials(db.vcap || { label: 'managed-hana' })

  if (!creds || typeof creds !== 'object' || !(creds.get_managed_instance_url || creds.sm_url)) {
    throw Object.assign(new Error('No or malformed Managed HANA credentials'), { credentials: creds })
  }

  // new instance manager
  return new Promise((resolve, reject) => {
    // REVISIT: better cache settings? current copied from old cds-hana...
    // note: may need to be low for mtx tests -> configurable?
    const opts = Object.assign(creds, {
      cache_max_items: 1,
      cache_item_expire_seconds: 1
    })
    _require('@sap/instance-manager').create(opts, (err, res) => {
      if (err) return reject(err)
      resolve(res)
    })
  })
}

function singleTenantInstanceManager (db = cds.env.requires.db) {
  const creds = db.credentials || _require('@sap/xsenv').serviceCredentials(db.vcap || { label: 'hana' })

  if (!creds || typeof creds !== 'object' || !creds.host) {
    throw Object.assign(new Error('No or malformed HANA credentials'), { credentials: creds })
  }

  // mock instance manager
  return {
    get: (_, cb) => {
      cb(null, { credentials: creds })
    }
  }
}

async function credentials4 (tenant, credentials) {
  if (!im) {
    const opts = credentials ? { credentials } : null
    im = cds.env.requires.db.multiTenant ? await multiTenantInstanceManager(opts) : singleTenantInstanceManager(opts)
  }

  return new Promise((resolve, reject) => {
    im.get(tenant, (err, res) => {
      if (err) return reject(err)
      if (!res) return reject(new Error(`There is no instance for tenant "${tenant}"`))
      resolve(res.credentials)
    })
  })
}

const addCheckServerIdentity = creds => {
  // REVISIT: copied from old cds-hana
  if (creds.sslValidateCertificate === false && creds.sslHostNameInCertificate) {
    const allowedHost = creds.sslHostNameInCertificate
    creds.checkServerIdentity = host => {
      if (host !== allowedHost) {
        throw new Error(
          `The name on the security certificate "${allowedHost}" is invalid or does not match the name of the site "${host}".`
        )
      }
    }
  }
}

const _ensureError = err => (err instanceof Error ? err : Object.assign(new Error(err.message), err))

const _connectHdb = creds => {
  addCheckServerIdentity(creds)

  return new Promise((resolve, reject) => {
    const client = hana.createClient(creds)
    client.name = hana.name // TODO find better way?
    client.connect(err => {
      if (err) reject(_ensureError(err))
      else {
        if (creds.schema) {
          client.exec(`SET SCHEMA ${creds.schema}`, err => {
            if (err) reject(Object.assign(new Error('Could not set schema'), err))
            else resolve(client)
          })
        } else resolve(client)
      }
    })
  })
}

const _connectHanaClient = creds => {
  return new Promise((resolve, reject) => {
    const client = hana.createConnection()
    client.name = hana.name

    // REVISIT
    creds.CURRENTSCHEMA = creds.schema

    client.connect(creds, err => {
      if (err) reject(_ensureError(err))
      else resolve(client)
    })
  })
}

function factory4 (creds) {
  return {
    create: function () {
      if (hana.name === 'hdb') {
        return _connectHdb(creds)
      }
      return _connectHanaClient(creds)
    },
    destroy: function (client) {
      return new Promise(resolve => {
        client.disconnect(err => {
          if (err) resolve()
          // REVISIT: what to do? ignore? crash app?
          else resolve()
        })
      })
    }
    /*
    validate: function(client) {
      return new Promise(async (resolve) => {
        resolve(await <stillValidCheckThatReturnsABoolean>())
      })
    }
    */
  }
}

const config = {
  min: 1,
  max: 100,
  evictionRunIntervalMillisForPools: 60 * 1000, // > REVISIT: what's this?!
  evictionRunIntervalMillis: 30 * 1000,
  acquireTimeoutMillis: 20 * 1000,
  numTestsPerEvictionRun: 2,
  softIdleTimeoutMillis: 30 * 1000,
  idleTimeoutMillis: 8 * 60 * 1000
}

// REVISIT: copied from old cds-hana
const _getMassagedCreds = function (creds) {
  if (!('ca' in creds) && creds.certificate) {
    creds.ca = creds.certificate
  }
  if ('encrypt' in creds && !('useTLS' in creds)) {
    creds.useTLS = creds.encrypt
  }
  if ('hostname_in_certificate' in creds && !('sslHostNameInCertificate' in creds)) {
    creds.sslHostNameInCertificate = creds.hostname_in_certificate
  }
  if ('validate_certificate' in creds && !('sslValidateCertificate' in creds)) {
    creds.sslValidateCertificate = creds.validate_certificate
  }
  return creds
}

const _getPoolConfig = function () {
  // REVISIT: where get pool config from?
  const mergedConfig = Object.assign({}, config, cds.env.requires.db.pool)

  // REVISIT: Shouldn't this check for the config of the required service and not db in every case?
  if (cds.env.requires.db.pool && !('numTestsPerEvictionRun' in cds.env.requires.db.pool)) {
    mergedConfig.numTestsPerEvictionRun = mergedConfig.min > 2 ? Math.ceil(mergedConfig.min / 2) : 2
  }

  return mergedConfig
}

const pools = new Map()

async function pool4 (tenant, credentials) {
  if (!pools.get(tenant)) {
    pools.set(
      tenant,
      new Promise((resolve, reject) => {
        credentials4(tenant, credentials)
          .then(creds => {
            resolve(pool.createPool(factory4(_getMassagedCreds(creds)), _getPoolConfig()))
          })
          .catch(e => {
            // delete pools entry if fetching credentials failed
            pools.delete(tenant)
            reject(e)
          })
      }).then(p => {
        pools.set(tenant, p)
        return p
      })
    )
  }
  if ('then' in pools.get(tenant)) {
    pools.set(tenant, await pools.get(tenant))
  }

  return pools.get(tenant)
}

module.exports = {
  acquire: (tenant, credentials) => {
    return pool4(tenant, credentials).then(p => p.acquire())
  },
  release: client => {
    return pool4(client._tenant).then(p => p.release(client))
  },
  drain: async tenant => {
    if (!pools.get(tenant)) {
      return
    }
    const p = await pool4(tenant)
    pools.delete(tenant)
    await p.drain()
    await p.clear()
  }
}
