// @sap/cds-runtime/lib/hana/Service.js
const cds = global.cds || require('@sap/cds/lib')

const DatabaseService = require('../db/Service')
const pool = require('./pool')

/*
 * hana-specific handlers
 */
const localized = require('./localized')
const dateTime = require('./dateTime')

/*
 * hana-specific execution
 */
const execute = require('./execute')

/*
 * helpers
 */
const _setSessionContext = (dbc, property, value) => {
  if (dbc._connection) {
    // Works, but bad practise to access an internal scope
    dbc._connection.getClientInfo().setProperty(property, value)
  } else {
    dbc.setClientInfo(property, value)
  }
}

/*
 * the service
 */
module.exports = class HanaDatabase extends DatabaseService {
  constructor (...args) {
    super(...args)

    // REVISIT: official db api
    this._execute = execute

    // REVISIT: db api
    this._insert = this._queries.insert(execute.insert)
    this._read = this._queries.read(execute.select, execute.stream)
    this._update = this._queries.update(execute.update, execute.select)
    this._delete = this._queries.delete(execute.delete)
    this._run = this._queries.run(this._insert, this._read, this._update, this._delete, execute.cqn, execute.sql)
  }

  init () {
    /*
     * before
     */
    this._ensureOpen && this.before('*', this._ensureOpen)
    this._ensureModel && this.before('*', this._ensureModel)

    this.before(['CREATE', 'UPDATE'], '*', dateTime) // > has to run before rewrite

    this.before(['CREATE', 'UPDATE'], '*', this._keys)
    this.before(['CREATE', 'UPDATE'], '*', this._managed)
    this.before(['CREATE', 'UPDATE'], '*', this._virtual)
    this.before(['CREATE', 'READ', 'UPDATE', 'DELETE'], '*', this._rewrite)

    this.before('READ', '*', localized) // > has to run after rewrite

    this.before('CREATE', '*', this._integrity.notNull)
    // REVISIT: get data to be deleted for integrity check
    this.before('DELETE', '*', this._integrity.beforeDelete)

    /*
     * on
     */
    this.on('CREATE', '*', this._CREATE)
    this.on('READ', '*', this._READ)
    this.on('UPDATE', '*', this._UPDATE)
    this.on('DELETE', '*', this._DELETE)

    /*
     * after
     */
    // REVISIT: after phase runs in parallel -> side effects possible!
    if (cds.env.odata_x4) {
      this.after(['READ'], '*', this._structured)
      this.after(['READ'], '*', this._arrayed)
    }

    /*
     * tx
     */
    this.on('BEGIN', async function (req) {
      this.dbc = await this.acquire(req)
      this.dbc.setAutoCommit(false)

      // REVISIT: compat for continue with tx
      this._state = req.event

      return 'dummy'
    })

    // REVISIT: register only if needed?
    this.before('COMMIT', this._integrity.performCheck)

    this.on(['COMMIT', 'ROLLBACK'], function (req) {
      const that = this
      return new Promise(function (resolve, reject) {
        that.dbc[req.event.toLowerCase()](async function (err) {
          try {
            that.dbc.setAutoCommit(true)
            await that.release(that.dbc)
          } catch (e) {
            // REVISIT: what to do?
            return reject(e)
          }

          // REVISIT: compat for continue with tx
          that._state = req.event

          if (err) return reject(err)
          resolve('dummy')
        })
      })
    })

    /*
     * generic
     */
    // all others, i.e. CREATE, DROP table, ...
    this.on('*', function (req) {
      return this._run(this.model, this.dbc, req.query || req.event, req, req.data)
    })
  }

  /*
   * connection
   */
  async acquire (arg) {
    const tenant = (typeof arg === 'string' ? arg : arg.user.tenant) || 'anonymous'
    const dbc = await pool.acquire(tenant, this.options.credentials)

    if (typeof arg !== 'string') {
      _setSessionContext(dbc, 'APPLICATIONUSER', arg.user.id || 'ANONYMOUS')
      _setSessionContext(dbc, 'LOCALE', arg.user.locale || 'en_US')
      if (arg._) {
        arg._['VALID-FROM'] && _setSessionContext(dbc, 'VALID-FROM', arg._['VALID-FROM'])
        arg._['VALID-TO'] && _setSessionContext(dbc, 'VALID-TO', arg._['VALID-TO'])
      }
    }

    dbc._tenant = tenant

    return dbc
  }

  release (dbc) {
    return pool.release(dbc)
  }

  // REVISIT: should happen automatically after a configurable time
  async disconnect (tenant = 'anonymous') {
    await pool.drain(tenant)
    super.disconnect(tenant)
  }
}
