// @sap/cds-runtime/lib/messaging/message-queuing-utils/management.js
const authorizedRequest = require('../common-utils/authorizedRequest')

const _oa2 = oauth2 => ({
  clientid: oauth2.clientId,
  clientsecret: oauth2.clientSecret,
  tokenendpoint: oauth2.tokenUrl
})
const addSubscription = (queueName, topicPattern, optionsClient, token) =>
  authorizedRequest({
    method: 'PUT',
    uri: optionsClient.management.url,
    path: `/v1/management/queues/${encodeURIComponent(queueName)}/subscriptions/topics/${encodeURIComponent(
      topicPattern
    )}`,
    oa2: _oa2(optionsClient.management.auth.oauth2),
    attemptInfo: () => console.log('[cds] - Add subscription', { topic: topicPattern, queue: queueName }),
    rejectString: `Subscription "${topicPattern}" could not be added to queue "${queueName}".`,
    token
  })

const putQueue = (queueName, optionsClient, options, token) =>
  authorizedRequest({
    method: 'PUT',
    uri: optionsClient.management.url,
    path: `/v1/management/queues/${encodeURIComponent(queueName)}`,
    oa2: _oa2(optionsClient.management.auth.oauth2),
    dataObj: options && options.queueConfig,
    attemptInfo: () => console.log('[cds] - Put queue', { queue: queueName }),
    rejectString: `Queue "${queueName}" could not be created.`,
    token
  })

const deleteQueue = (queueName, optionsClient, token) =>
  authorizedRequest({
    method: 'DELETE',
    uri: optionsClient.management.url,
    path: `/v1/management/queues/${encodeURIComponent(queueName)}`,
    oa2: _oa2(optionsClient.management.auth.oauth2),
    attemptInfo: () => console.log('[cds] - Delete queue', { queue: queueName }),
    rejectString: `Queue "${queueName}" could not be deleted.`,
    token
  })

module.exports = {
  addSubscription,
  putQueue,
  deleteQueue
}
