// @sap/cds-runtime/lib/cds-services/services/handlers/onCreateDraft.js
const onDraftActivate = require('./onDraftActivate')._handler

const { isNavigationToMany } = require('../utils/compositionTree')
const { removeDraftUUID, ensureDraftsSuffix } = require('../utils/draftUtils')
const { DRAFT_COLUMNS } = require('../utils/constants')

const _getUpdateDraftAdminCQN = ({ statements, user }, draftUUID) => {
  return statements
    .UPDATE('DRAFT.DraftAdministrativeData')
    .data({
      InProcessByUser: user.id,
      LastChangedByUser: user.id,
      LastChangeDateTime: new Date().toISOString()
    })
    .where({ DraftUUID: draftUUID })
}

const _getInsertDraftAdminCQN = ({ statements, user }, uuid) => {
  const time = new Date().toISOString()

  return statements.INSERT.into('DRAFT.DraftAdministrativeData').entries({
    DraftUUID: uuid,
    CreationDateTime: time,
    CreatedByUser: user.id,
    LastChangeDateTime: time,
    LastChangedByUser: user.id,
    DraftIsCreatedByMe: true,
    DraftIsProcessedByMe: true,
    InProcessByUser: user.id
  })
}

const _getInsertDataCQN = (context, draftUUID) => {
  const draftName = ensureDraftsSuffix(context.target.name)

  const insertData = context.statements.INSERT.into(draftName).entries(context.query.INSERT.entries[0]) // entries is always set because there are no entities without keys

  context.data.IsActiveEntity = false
  context.data.HasDraftEntity = false
  context.data.HasActiveEntity = false
  context.data.DraftAdministrativeData_DraftUUID = draftUUID

  return insertData
}

/**
 * Generic Handler for CREATE requests in the context of draft.
 * In case of success it returns the created entry.
 *
 * @alias module:handlers.onCreateDraft
 */
const _handler = (req, next) => {
  if (!req._draftMetadata) {
    // REVISIT: when is this the case?
    return onDraftActivate(req, next)
  }

  // fill default values
  const elements = req.target.elements
  for (const column of Object.keys(elements)) {
    const col = elements[column]
    if (col.default !== undefined && !DRAFT_COLUMNS.includes(column)) {
      req.data[col.name] = 'val' in col.default ? col.default.val : col.default
    }
  }

  const navigationToMany = isNavigationToMany(req)

  const adminDataCQN = navigationToMany
    ? _getUpdateDraftAdminCQN(req, req.data.DraftAdministrativeData_DraftUUID)
    : _getInsertDraftAdminCQN(req, req.data.DraftAdministrativeData_DraftUUID)
  const insertDataCQN = _getInsertDataCQN(req, req.data.DraftAdministrativeData_DraftUUID)

  // read data as on db and return
  const keyName = Object.values(req.target.elements).find(val => val.key).name
  const readInsertDataCQN = req.statements.SELECT.from(insertDataCQN.INSERT.into).where(keyName, '=', req.data[keyName])

  return Promise.all([req.run(adminDataCQN), req.run(insertDataCQN)])
    .then(() => {
      return req.run(readInsertDataCQN)
    })
    .then(result => {
      if (result.length === 0) {
        req.reject(404)
      }
      return removeDraftUUID(result[0])
    })
}

const { ODATA, COMMON } = require('../../../common/constants/annotation')
const _relevant = e => e[ODATA.DRAFT] || e[COMMON.DRAFT_NODE.PREP_ACTION]
module.exports = function () {
  for (const entity of Object.values(this.entities).filter(_relevant)) {
    this.on('NEW', entity, _handler)
  }
}
