// @sap/cds-runtime/lib/db/Service.js
const cds = global.cds || require('@sap/cds/lib')

const { Transform } = require('stream')

/*
 * generic queries
 */
const queries = require('./query')

/*
 * generic handlers
 */
const generic = require('./generic')

/*
 * helpers
 */
const _isOpen = {
  new: true, // > REVISIT: nothing done on transaction, but we still need to commit/rollback due to this._busy
  BEGIN: true
}

module.exports = class DatabaseService extends cds.Service {
  constructor (...args) {
    super(...args)

    // REVISIT: official db api
    this._queries = queries

    // REVISIT: official db api
    for (const each in generic) {
      this[`_${each}`] = generic[each]
    }

    // REVISIT: compat for continue with tx, but client should not be released on commit/rollback
    this._ensureOpen = async function (req) {
      if (req.event === 'BEGIN') return
      if (this._state && this._state !== 'BEGIN') await this.begin()
    }
    this._ensureOpen._initial = true

    // REVISIT: ensures there is an this.model if this is a transaction
    this._ensureModel = function (req) {
      if (this._is_tx && !this.model) this.model = req._model
    }
    this._ensureModel._initial = true

    // REVISIT: how to generic handler registration?
  }

  set model (m) {
    // Ensure the model we get has unfolded entities for localized data, drafts, etc.
    // Note: cds.deploy and some tests set the model of cds.db outside the constructor
    super.model = m && 'definitions' in m ? cds.compile.for.odata(m) : m
  }

  /*
   * tx
   */
  begin () {
    return this.emit('BEGIN')
  }

  commit () {
    // REVISIT: move to cds.Service?
    if (_isOpen[this._state]) {
      return this.emit('COMMIT')
    }

    // REVISIT: should not be necessary
    return Promise.resolve('dummy') // > nothing to do
  }

  rollback () {
    // REVISIT: move to cds.Service?
    if (_isOpen[this._state]) {
      return this.emit('ROLLBACK')
    }

    // REVISIT: should not be necessary
    return Promise.resolve('dummy') // > nothing to do
  }

  /*
   * streaming
   */
  _runStream (streamQuery, result) {
    this.run(streamQuery).then(stream => {
      if (!stream) {
        result.push(null)
      } else {
        stream.value.pipe(result)
      }
    })
  }

  stream (query) {
    // aynchronous API: cds.stream(query)
    if (typeof query === 'object') {
      return new Promise(async (resolve, reject) => {
        try {
          const res = await this.run(Object.assign(query, { _streaming: true }))
          resolve((res && res.value) || res)
        } catch (e) {
          reject(e)
        }
      })
    }

    // synchronous API: cds.stream('column').from(entity).where(...)
    return {
      from: (...args) => {
        const streamQuery = cds.ql.SELECT.from(...args)
        if (!streamQuery.SELECT.columns || streamQuery.SELECT.columns.length !== 0) {
          streamQuery.columns([query])
        }
        delete streamQuery.SELECT.one
        streamQuery._streaming = true

        const result = new Transform({
          transform (chunk, encoding, callback) {
            this.push(chunk)
            callback()
          }
        })

        if (!streamQuery.SELECT.where) {
          return {
            where: (...args) => {
              streamQuery.where(...args)
              this._runStream(streamQuery, result)

              return result
            }
          }
        }

        this._runStream(streamQuery, result)

        return result
      }
    }
  }
}
