// @sap/cds-dk/bin/activate.js
module.exports = Object.assign(activate, {
    options: ['--subdomain', '--passcode', '--username', '--to'],
    flags: ['--undeploy', '--wsdl'],
    shortcuts: ['-s', '-p', '-u'],
    help: `
# SYNOPSIS

    *cds activate* [ <project> ]

   Activate an extension project


# OPTIONS

    *--to* <url>
        The url to activate.

    *-s* | *--subdomain* <subdomain>

        The subdomain to activate to.

    *-p* | *--passcode* <passcode>

        The passcode for authentication

    *-u* | *--username* <name>

        Deprecated. Use passcode parameter instead.

    *--undeploy*
        Undeploy a previously deployed extension.

    *--wsdl*
        Upload files from wsdl folder

`})


async function activate ([projectFolder], options = {}) {
    const injection = {
        cds: require('@sap/cds/lib')
    }

    Object.assign(options, { projectFolder })

    const client = require('@sap/cds-sidecar-client')
    await client.apply(injection, options)
}



/* eslint no-console: off */
