// @sap/cds-runtime/lib/sqlite/Service.js
const cds = global.cds || require('@sap/cds/lib')

const DatabaseService = require('../db/Service')
const sqlite = require('sqlite3')

/*
 * sqlite-specific handlers
 */
const localized = require('./localized')

/*
 * sqlite-specific execution
 */
const execute = require('./execute')

/*
 * helpers
 */
const colored = {
  BEGIN: '\x1b[1m\x1b[33mBEGIN\x1b[0m',
  COMMIT: '\x1b[1m\x1b[32mCOMMIT\x1b[0m',
  ROLLBACK: '\x1b[1m\x1b[91mROLLBACK\x1b[0m'
}

const _new = url => {
  return new Promise((resolve, reject) => {
    const dbc = new sqlite.Database(url, err => {
      if (err) return reject(err)

      // REVISIT: debug logging
      const { DEBUG } = process.env
      if (/\b(y|sqlite)\b/.test(DEBUG)) {
        dbc.on('trace', sql => {
          return console.debug(colored[sql] || sql)
        })
      }

      resolve(dbc)
    })
  })
}

/*
 * the service
 */
module.exports = class SQLiteDatabase extends DatabaseService {
  constructor (...args) {
    super(...args)

    // REVISIT: official db api
    this._execute = execute

    // REVISIT: official db api
    this._insert = this._queries.insert(execute.insert)
    this._read = this._queries.read(execute.select, execute.stream)
    this._update = this._queries.update(execute.update, execute.select)
    this._delete = this._queries.delete(execute.delete)
    this._run = this._queries.run(this._insert, this._read, this._update, this._delete, execute.cqn, execute.sql)

    this.dbcs = new Map()
  }

  set model (csn) {
    const m = csn && 'definitions' in csn ? cds.linked(cds.compile.for.odata(csn)) : csn
    cds.alpha_localized(m)
    super.model = m
  }

  init () {
    /*
     * before
     */
    this._ensureOpen && this.before('*', this._ensureOpen)
    this._ensureModel && this.before('*', this._ensureModel)

    this.before(['CREATE', 'UPDATE'], '*', this._keys)
    this.before(['CREATE', 'UPDATE'], '*', this._managed)
    this.before(['CREATE', 'UPDATE'], '*', this._virtual)
    this.before(['CREATE', 'READ', 'UPDATE', 'DELETE'], '*', this._rewrite)

    this.before('READ', '*', localized) // > has to run after rewrite

    this.before('CREATE', '*', this._integrity.notNull)
    // REVISIT: get data to be deleted for integrity check
    this.before('DELETE', '*', this._integrity.beforeDelete)

    /*
     * on
     */
    this.on('CREATE', '*', this._CREATE)
    this.on('READ', '*', this._READ)
    this.on('UPDATE', '*', this._UPDATE)
    this.on('DELETE', '*', this._DELETE)

    /*
     * after
     */
    // REVISIT: after phase runs in parallel -> side effects possible!
    if (cds.env.odata_x4) {
      // REVISIT only register for entities that contain structured/arrayed or navigation to it
      this.after(['READ'], '*', this._structured)
      this.after(['READ'], '*', this._arrayed)
    }

    /*
     * tx
     */
    this.on('BEGIN', async function (req) {
      this.dbc = await this.acquire(req)
      const res = this._run(this.model, this.dbc, req.event)

      // REVISIT: compat for continue with tx
      this._state = req.event

      return res
    })

    // REVISIT: register only if needed?
    this.before('COMMIT', this._integrity.performCheck)

    this.on(['COMMIT', 'ROLLBACK'], async function (req) {
      const res = await this._run(this.model, this.dbc, req.event)

      // REVISIT: compat for continue with tx
      this._state = req.event

      this.release(this.dbc)
      return res
    })

    /*
     * generic
     */
    // all others, i.e. CREATE, DROP table, ...
    this.on('*', function (req) {
      return this._run(this.model, this.dbc, req.query || req.event, req)
    })
  }

  /*
   * connection
   */
  async acquire (arg) {
    const tenant = (typeof arg === 'string' ? arg : arg.user.tenant) || 'anonymous'

    let dbc = this.dbcs.get(tenant)
    if (!dbc) {
      const credentials = this.options.credentials || this.options || {}
      let dbUrl = credentials.database || credentials.url || credentials.host || ':memory:'

      if (this.options.multiTenant && dbUrl.endsWith('.db')) {
        dbUrl = dbUrl.split('.db')[0] + '_' + tenant + '.db'
      }

      dbc = await _new(dbUrl)

      dbc._queued = []
      dbc._tenant = tenant

      this.dbcs.set(tenant, dbc)
    }

    if (dbc._busy) await new Promise(resolve => dbc._queued.push(resolve))
    else dbc._busy = true

    return dbc
  }

  release (dbc) {
    if (dbc._queued.length) dbc._queued.shift()()
    else dbc._busy = false
  }

  /*
   * deploy
   */
  // REVISIT: make tenant aware
  async deploy (model, options = {}) {
    const createEntities = cds.compile.to.sql(model)
    if (!createEntities || createEntities.length === 0) return // > nothing to deploy

    const dropViews = []
    const dropTables = []
    for (let each of createEntities) {
      const [, table, entity] = each.match(/^\s*CREATE (?:(TABLE)|VIEW)\s+"?([^\s(]+)"?/im) || []
      if (table) dropTables.push({ DROP: { entity } })
      else dropViews.push({ DROP: { view: entity } })
    }

    if (options.dry) {
      const log = console.log // eslint-disable-line no-console
      for (let {
        DROP: { view }
      } of dropViews) {
        log('DROP VIEW IF EXISTS ' + view + ';')
      }
      log()
      for (let {
        DROP: { entity }
      } of dropTables) {
        log('DROP TABLE IF EXISTS ' + entity + ';')
      }
      log()
      for (let each of createEntities) log(each + ';\n')
      return
    }

    const tx = this.transaction()
    await tx.run(dropViews)
    await tx.run(dropTables)
    await tx.run(createEntities)
    await tx.commit()

    return true
  }
}
