// @sap/cds-runtime/lib/messaging/enterprise-messaging-utils/management.js
const authorizedRequest = require('../common-utils/authorizedRequest')

const putQueue = (queueName, optionsClient, options, token) =>
  authorizedRequest({
    method: 'PUT',
    uri: optionsClient.management.uri,
    path: `/hub/rest/api/v1/management/messaging/queues/${encodeURIComponent(queueName)}`,
    oa2: optionsClient.management.oa2,
    dataObj: options && options.queueConfig,
    attemptInfo: () => console.log('[cds] - Put queue', { queue: queueName }),
    rejectString: `Queue "${queueName}" could not be created.`,
    token
  })

const addSubscription = (queueName, topicPattern, optionsClient, token) =>
  authorizedRequest({
    method: 'PUT',
    uri: optionsClient.management.uri,
    path: `/hub/rest/api/v1/management/messaging/queues/${encodeURIComponent(
      queueName
    )}/subscriptions/${encodeURIComponent(topicPattern)}`,
    oa2: optionsClient.management.oa2,
    attemptInfo: () => console.log('[cds] - Add subscription', { topic: topicPattern, queue: queueName }),
    rejectString: `Subscription "${topicPattern}" could not be added to queue "${queueName}".`,
    token
  })

const deleteQueue = (queueName, optionsClient, token) =>
  authorizedRequest({
    method: 'DELETE',
    uri: optionsClient.management.uri,
    path: `/hub/rest/api/v1/management/messaging/queues/${encodeURIComponent(queueName)}`,
    oa2: optionsClient.management.oa2,
    attemptInfo: () => console.log('[cds] - Delete queue', { queue: queueName }),
    rejectString: `Queue "${queueName}" could not be deleted.`,
    token
  })

module.exports = {
  addSubscription,
  putQueue,
  deleteQueue
}
