// @sap-cloud-sdk/core/dist/odata/v2/uri-conversion/odata-uri.js
"use strict";
/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
Object.defineProperty(exports, "__esModule", { value: true });
exports.oDataUri = void 0;
var common_1 = require("../../common");
var get_expand_1 = require("./get-expand");
var get_select_1 = require("./get-select");
var uri_value_converter_1 = require("./uri-value-converter");
/**
 * @experimental This is experimental and is subject to change. Use with caution.
 */
exports.oDataUri = {
    getExpand: get_expand_1.getExpand,
    getFilter: common_1.createGetFilter({ convertToUriFormat: uri_value_converter_1.convertToUriFormat }).getFilter,
    getEntityKeys: common_1.getEntityKeys,
    getOrderBy: common_1.getOrderBy,
    getResourcePathForKeys: function (keys, entityConstructor) {
        if (keys === void 0) { keys = {}; }
        return common_1.createGetResourcePathForKeys({
            convertToUriFormat: uri_value_converter_1.convertToUriFormat
        }).getResourcePathForKeys(keys, entityConstructor);
    },
    getSelect: get_select_1.getSelect,
    convertToUriFormat: uri_value_converter_1.convertToUriFormat
};
