// @sap-cloud-sdk/core/dist/odata/common/selectable/one-to-many-link.js
"use strict";
/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.OneToManyLink = void 0;
var filter_1 = require("../filter");
var link_1 = require("./link");
/**
 * @experimental
 */
var OneToManyLink = /** @class */ (function (_super) {
    __extends(OneToManyLink, _super);
    function OneToManyLink() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._orderBy = [];
        return _this;
    }
    OneToManyLink.prototype.clone = function () {
        var clonedLink = _super.prototype.clone.call(this);
        clonedLink._filters = this._filters;
        clonedLink._orderBy = this._orderBy;
        clonedLink._top = this._top;
        clonedLink._skip = this._skip;
        return clonedLink;
    };
    // eslint-disable-next-line valid-jsdoc
    /**
     * @experimental This is experimental and is subject to change. Use with caution.
     */
    OneToManyLink.prototype.filter = function () {
        var expressions = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            expressions[_i] = arguments[_i];
        }
        var link = this.clone();
        link._filters = new filter_1.FilterLink(this, expressions.map(function (e) { return (e instanceof OneToManyLink ? e._filters : e); }));
        return link;
    };
    /**
     * Add order-by statements to the request.
     *
     * @param orderBy - OrderBy statements to order the response by
     * @returns The request builder itself, to facilitate method chaining
     */
    OneToManyLink.prototype.orderBy = function () {
        var orderBy = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            orderBy[_i] = arguments[_i];
        }
        var link = this.clone();
        link._orderBy = orderBy;
        return link;
    };
    /**
     * Limit number of returned entities.
     *
     * @param top - Maximum number of entities to return in the response. Can be less, if less entities match the request
     * @returns The request builder itself, to facilitate method chaining
     */
    OneToManyLink.prototype.top = function (top) {
        var link = this.clone();
        link._top = top;
        return link;
    };
    /**
     * Skip number of entities.
     *
     * @param skip - Number of matching entities to skip. Useful for paging
     * @returns The request builder itself, to facilitate method chaining
     */
    OneToManyLink.prototype.skip = function (skip) {
        var link = this.clone();
        link._skip = skip;
        return link;
    };
    return OneToManyLink;
}(link_1.Link));
exports.OneToManyLink = OneToManyLink;
