// @sap-cloud-sdk/core/dist/odata/v4/request-builder/response-transformers.js
"use strict";
/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformReturnValueForEdmTypeList = exports.transformReturnValueForEdmType = exports.transformReturnValueForComplexTypeList = exports.transformReturnValueForComplexType = exports.transformReturnValueForEntityList = exports.transformReturnValueForEntity = exports.transformReturnValueForUndefined = void 0;
var entity_deserializer_1 = require("../entity-deserializer");
var response_data_accessor_1 = require("./response-data-accessor");
function transformReturnValueForUndefined(data, builderFn) {
    return builderFn(data);
}
exports.transformReturnValueForUndefined = transformReturnValueForUndefined;
function transformReturnValueForEntity(data, entityConstructor) {
    return entity_deserializer_1.deserializeEntity(response_data_accessor_1.getSingleResult(data), entityConstructor).setOrInitializeRemoteState();
}
exports.transformReturnValueForEntity = transformReturnValueForEntity;
function transformReturnValueForEntityList(data, entityConstructor) {
    return response_data_accessor_1.getCollectionResult(data).map(function (entityJson) {
        return entity_deserializer_1.deserializeEntity(entityJson, entityConstructor).setOrInitializeRemoteState();
    });
}
exports.transformReturnValueForEntityList = transformReturnValueForEntityList;
function transformReturnValueForComplexType(data, builderFn) {
    return builderFn(response_data_accessor_1.getSingleResult(data));
}
exports.transformReturnValueForComplexType = transformReturnValueForComplexType;
function transformReturnValueForComplexTypeList(data, builderFn) {
    return response_data_accessor_1.getCollectionResult(data).map(function (json) { return builderFn(json); });
}
exports.transformReturnValueForComplexTypeList = transformReturnValueForComplexTypeList;
function transformReturnValueForEdmType(data, builderFn) {
    return builderFn(response_data_accessor_1.getSingleResult(data));
}
exports.transformReturnValueForEdmType = transformReturnValueForEdmType;
function transformReturnValueForEdmTypeList(data, builderFn) {
    return response_data_accessor_1.getCollectionResult(data).map(builderFn);
}
exports.transformReturnValueForEdmTypeList = transformReturnValueForEdmTypeList;
